### PR TITLE
syslog: Added multi device support in syslog_device.

### PR DESCRIPTION
**Pull request by fjpanag**
## Summary
Added support for multiple devices in syslog_device.

Continuing my work from PR #3050, we can now have multiple instances of a syslog_device, so the different channels can co-exist correctly.

This opens up the way for future development like adding/removing channels dynamically, implementing different thread locking of the syslog etc etc.

## Impact
It shouldn't have any impact on any existing configuration.

## Testing
For the time being I have only done some minimal testing (mostly that it builds OK).  
I plan to do more testing (and will confirm bellow in a comment).  

The PR is open however, to allow for reviewing while I test.

### CODE REVIEW

File: drivers/syslog/syslog.h
Comment (gustavonihei) [inline drivers/syslog/syslog.h:93]:
```suggestion
 *   oflags     - File open flags.
 *   mode       - File open mode (only if oflags include O_CREAT).
```
nit: Just for consistency.

File: drivers/syslog/syslog.h
Comment (gustavonihei) [inline drivers/syslog/syslog.h:314]:
```suggestion
 *   buffer     - The buffer containing the data to be output.
 *   buflen     - The number of bytes in the buffer.
```
nit: Just for consistency.

File: drivers/syslog/syslog_device.c
Comment (gustavonihei) [inline drivers/syslog/syslog_device.c:436]:
```suggestion
 *   buffer     - The buffer containing the data to be output.
 *   buflen     - The number of bytes in the buffer.
```

File: drivers/syslog/syslog.h
Comment (fjpanag) [reply drivers/syslog/syslog.h]: Done.

File: arch/arm/src/armv7-m/arm_itm_syslog.c
Comment (gustavonihei) [inline arch/arm/src/armv7-m/arm_itm_syslog.c:112]:
```suggestion
  UNUSED(dev);
```
You may use the UNUSED macro from `nuttx/compiler.h`.

File: drivers/syslog/syslog.h
Comment (fjpanag) [reply drivers/syslog/syslog.h]: done

File: include/nuttx/syslog/syslog.h
Comment (gustavonihei) [inline include/nuttx/syslog/syslog.h:96]:
```suggestion
  FAR void *sc_dev;
```
Missing FAR qualifier on common code.

File: drivers/syslog/syslog_device.c
Comment (fjpanag) [reply drivers/syslog/syslog_device.c]: done

File: drivers/syslog/syslog_channel.c
Comment (v01d) [inline drivers/syslog/syslog_channel.c:49]: I'm confused about this change in this PR.

Comment (fjpanag) [reply drivers/syslog/syslog_channel.c:49]: The function `sc_putc()` has changed prototype, so a pointer to the device is passed:
```c
typedef CODE int (*syslog_putc_t)(FAR void *dev, int ch);
```

Thus `up_putc()` is incompatible. So I use `syslog_default_putc()` to translate the arguments from the new `sc_putc()` to the existing `up_putc()`.

Notice the new argument in `drivers/syslog/syslog_channel.c`, line 112.

File: include/nuttx/syslog/syslog.h
Comment (fjpanag) [reply include/nuttx/syslog/syslog.h]: I'm sorry, I always forget `FAR`s... Added now.

Comment (gustavonihei) [reply include/nuttx/syslog/syslog.h]: No problem!

File: drivers/syslog/syslog_channel.c
Comment (v01d) [reply drivers/syslog/syslog_channel.c:49]: Got it, much clearer now

File: arch/arm/src/armv7-m/arm_itm_syslog.c
Comment (fjpanag) [reply arch/arm/src/armv7-m/arm_itm_syslog.c:112]: I think this is not covered in coding standard.  
What is the preferred style in NuttX?

Also, maybe someone to add this in coding standard?

Comment (gustavonihei) [reply arch/arm/src/armv7-m/arm_itm_syslog.c:112]: I believe this is something that should not be covered in the NuttX Style Guide.
It is more of a code quality question, to make the code cleaner.

File: drivers/syslog/syslog_channel.c
Comment (fjpanag) [reply drivers/syslog/syslog_channel.c:49]: Just to note, my intention is to add a new channel in the future, that it will encapsulate ARM ITM, up_putc, rpmsg etc, so even these interfaces follow the common structure, and use a proper `struct syslog_dev_s` instead of this ungly `void *` in `struct syslog_channel_s`.

I believe that it will simplify the code more.

Comment (v01d) [reply drivers/syslog/syslog_channel.c:49]: I agree, that is a good goal

File: arch/arm/src/armv7-m/arm_itm_syslog.c
Comment (fjpanag) [reply arch/arm/src/armv7-m/arm_itm_syslog.c:112]: AFAIK `(void) unused_variable;` is perfectly valid in C, with no portability issues.  
Or if I am wrong, please enlighten me.

So this is why I consider this a "styling" issue.

Nevertheless, I will update all occurrences of this to use the `UNUSED()` macro. :)

Comment (v01d) [reply arch/arm/src/armv7-m/arm_itm_syslog.c:112]: It is a matter of consistency with the rest of the codebase. Also, I think the macro is more verbose in its purpose, that's why it is used.

File: drivers/syslog/syslog_channel.c
Comment (xiaoxiang781216) [reply drivers/syslog/syslog_channel.c:49]: @fjpanag it looks that syslog channel become a normal driver. Should we follow other drivers convention? e.g. spi_dev_s/spi_ops_s or rtc_lowerhalf_s/rtc_ops_s, so other people will be more familiar with your change.

File: arch/arm/src/armv7-m/arm_itm_syslog.c
Comment (fjpanag) [reply arch/arm/src/armv7-m/arm_itm_syslog.c]: I think it is fixed everywhere now.

File: drivers/syslog/syslog_channel.c
Comment (fjpanag) [reply drivers/syslog/syslog_channel.c:50]: Yes, this is what I had in mind, but you expressed it in a better way. Having upper/lower drivers for syslog too.

However this means that new lower drivers are to be created (as mentioned above), files may need to be moved etc.

I would prefer to finalize this migration in a later PR, after the lower-hald driver interface is finalized (`sc_close()` is still to be introduced, thread-safety protections are to be added and more).

What do you think? Shall we keep this part for the moment?

Comment (fjpanag) [reply drivers/syslog/syslog_channel.c:50]: Something that I am also thinking is to completely drop `sc_putc()` and require on every lower-half driver to implement an `sc_write()` only, with the requirement that the driver will perform the write atomically. If the driver cannot actually perform a write, it can iterate internally in a protection section of the code.

Comment (xiaoxiang781216) [reply drivers/syslog/syslog_channel.c:50]: > Yes, this is what I had in mind, but you expressed it in a better way. Having upper/lower drivers for syslog too.
> 
> However this means that new lower drivers are to be created (as mentioned above), files may need to be moved etc.
> 

Not to much to become upper/lower:

1. Rename syslog_channel_s to syslog_channel_ops_s
2. Add new struct syslog_channel_s which contain the pointer of syslog_channel_ops_s
3. The callback in syslog_channel_ops_s accept the pointer of syslog_channel_s as the new first argument

> I would prefer to finalize this migration in a later PR, after the lower-half driver interface is finalized (`sc_close()` is still to be introduced, thread-safety protections are to be added and more).
> 

The more callback like sc_close could be added in later if required

> What do you think? Shall we keep this part for the moment?

I am fine which one come first, my major concern is to reduce your implmentation cost.

> Something that I am also thinking is to completely drop `sc_putc()` and require on every lower-half driver to implement an `sc_write()` only, with the requirement that the driver will perform the write atomically. If the driver cannot actually perform a write, it can iterate internally in a protected section of the code.

Yes, I also think sc_putc make the interface more complicated than necessary. Maybe, it's better to remove sc_putc first before we redesign the syslog channel.

**Comment by fjpanag** [comment]
Regarding the style issue:
```
/home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/drivers/syslog/syslog_write.c:100:83: error: Long line found
```
has anyone any recommendation how this is to be fixed?

File: include/nuttx/syslog/syslog.h
Comment (xiaoxiang781216) [inline include/nuttx/syslog/syslog.h:87]: should we pass ```FAR struct syslog_channel_s *``` instead

Comment (fjpanag) [reply include/nuttx/syslog/syslog.h:87]: I was thinking that it is better to keep it as a `void *` temporarily, as the underlying channels are not uniform yet (like ARM ITM for example).

But I think I will just do it properly, and go for your your recommendation above.

File: drivers/syslog/syslog_channel.c
Comment (fjpanag) [reply drivers/syslog/syslog_channel.c:50]: @xiaoxiang781216 If I understand this correctly, `struct syslog_dev_s` should be moved to `include/nuttx/syslog/syslog.h`.
And then `struct syslog_channel_s` will be something like:

```c
struct syslog_channel_s
{
  /* Channel device */

  FAR struct syslog_dev_s *sc_dev;

  /* Channel operations */

  FAR struct syslog_channel_ops_s *ops;

  /* Implementation specific logic may follow */
};
```

Correct?
Or should `sc_dev` remain a `void *`, and keep `struct syslog_dev_s *sc_dev` private?

File: include/nuttx/syslog/syslog.h
Comment (xiaoxiang781216) [reply include/nuttx/syslog/syslog.h:87]: My suggestion come from other drivers(e.g. spi_dev_s/spi_ops_s) practice. It's better to do the right interface design in the first step, instead changing the interfae prootype again and again.

Comment (fjpanag) [reply include/nuttx/syslog/syslog.h]: @xiaoxiang781216 You may have a look, it seems better now.

I think this is what you had in mind.

File: include/nuttx/syslog/syslog.h
Comment (fjpanag) [reply include/nuttx/syslog/syslog.h]: done

File: include/nuttx/syslog/syslog.h
Comment (fjpanag) [reply include/nuttx/syslog/syslog.h:92]: Why is that a problem? Isn't it constant?

If const is to be removed, then all channels will have to also be declared as non-const.  
Else they will have to be always be cast to non-const, in which case there is no meaning to the const declaration in the firt place.

Comment (xiaoxiang781216) [reply include/nuttx/syslog/syslog.h:92]: Because the callback normally will modify syslog_channel_s, but I am fine if you want to make it const

File: include/nuttx/syslog/syslog.h
Comment (xiaoxiang781216) [inline include/nuttx/syslog/syslog.h:119]: remove sc_dev, let's driver embed syslog_channel_s into his own struct directly like this:
```
struct my_channel_s
{
  struct syslog_channel_s;
 /* add more sate here */
};

static int my_syslog_putc)(FAR struct syslog_channel_s *channel,  int ch)
{
  FAR struct my_channel_s *my = (FAR struct my_channel_s *)channel;

  /* access the state by my->xxx */
}
```
This is the convention used by many driver.

File: drivers/syslog/syslog_filechannel.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_filechannel.c:71]: remove, let syslog_dev_initialize return ```FAR syslog_channel_s *``` instead

File: drivers/syslog/syslog_devchannel.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_devchannel.c:94]: remove, let syslog_dev_initialize return ```FAR syslog_channel_s *``` instead

File: drivers/syslog/syslog_chardev.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_chardev.c:33]: remove, don't need

File: drivers/syslog/syslog_consolechannel.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_consolechannel.c:81]: remove, let syslog_dev_initialize return ```FAR syslog_channel_s *``` instead

File: include/nuttx/syslog/syslog.h
Comment (fjpanag) [reply include/nuttx/syslog/syslog.h]: @xiaoxiang781216 Please have a look whether it is OK now, so I can proceed with testing it.

File: drivers/syslog/syslog_device.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_device.c:134]:
```suggestion
static inline void syslog_dev_givesem(FAR struct syslog_dev_s *syslog_dev)
```

File: drivers/syslog/syslog_device.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_device.c:186]:
```suggestion
static int syslog_dev_outputready(FAR struct syslog_dev_s *syslog_dev)
```

File: drivers/syslog/syslog_device.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_device.c:93]:
```suggestion
static inline int syslog_dev_takesem(FAR struct syslog_dev_s *syslog_devl)
```

File: drivers/syslog/syslog.h
Comment (xiaoxiang781216) [inline drivers/syslog/syslog.h:114]: let syslog_dev_initialize allocate syslog_channel_s(syslog_dev_s) internally instead

File: drivers/syslog/syslog.h
Comment (xiaoxiang781216) [inline drivers/syslog/syslog.h:58]: move to syslog_device.c

File: drivers/syslog/syslog.h
Comment (xiaoxiang781216) [inline drivers/syslog/syslog.h:71]: move to syslog_device.c

File: drivers/syslog/syslog_device.c
Comment (xiaoxiang781216) [inline drivers/syslog/syslog_device.c:78]: Don't remove, let's change to:
```
* This structure contains all SYSLOGing state information */

struct syslog_dev_s
{
  struct syslog_channel_s channel;
  uint8_t      sl_state;    /* See enum syslog_dev_state */
  uint8_t      sl_oflags;   /* Saved open mode (for re-open) */
  uint16_t     sl_mode;     /* Saved open flags (for re-open) */
  sem_t        sl_sem;      /* Enforces mutually exclusive access */
  pid_t        sl_holder;   /* PID of the thread that holds the semaphore */
  struct file  sl_file;     /* The syslog file structure */
  FAR char    *sl_devpath;  /* Full path to the character device */
};
```

File: drivers/syslog/syslog_chardev.c
Comment (fjpanag) [reply drivers/syslog/syslog_chardev.c]: Yep, removed. Was added by mistake.

File: drivers/syslog/syslog_filechannel.c
Comment (fjpanag) [reply drivers/syslog/syslog_filechannel.c]: done

File: drivers/syslog/syslog_devchannel.c
Comment (fjpanag) [reply drivers/syslog/syslog_devchannel.c]: done

File: drivers/syslog/syslog_consolechannel.c
Comment (fjpanag) [reply drivers/syslog/syslog_consolechannel.c]: done

File: drivers/syslog/syslog_device.c
Comment (fjpanag) [reply drivers/syslog/syslog_device.c]: done

File: drivers/syslog/syslog_device.c
Comment (fjpanag) [reply drivers/syslog/syslog_device.c]: done

File: drivers/syslog/syslog.h
Comment (fjpanag) [reply drivers/syslog/syslog.h]: done

File: drivers/syslog/syslog_device.c
Comment (fjpanag) [reply drivers/syslog/syslog_device.c]: done

File: drivers/syslog/syslog.h
Comment (fjpanag) [reply drivers/syslog/syslog.h]: Yes I followed this idea. It is indeed better now.

File: drivers/syslog/syslog.h
Comment (fjpanag) [reply drivers/syslog/syslog.h]: done

File: drivers/syslog/syslog_device.c
Comment (fjpanag) [reply drivers/syslog/syslog_device.c:78]: I did it like this. `struct syslog_dev_state_s` is removed and `struct syslog_dev_s` is now private.

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

**Comment by fjpanag** [comment]
> LGTM.

Please don't merge till I confirm finishing of any testing still pending.

**Comment by fjpanag** [comment]
@xiaoxiang781216 I finished with any tests that I had in mind, and I didn't had any problems with this.

Unfortunately, I am not able to test all possible configurations, but at least the serial and the file loggers (and the underlying syslog_device) work correctly.  
The rest of the changes seem to be pretty trivial.

I found a couple of issues, but they are not directly related to these changes, so I will proceed with two new PRs immediately after merging this.

Feel free to merge it, if you also think it is OK.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
Sure.

**Comment by v01d** [comment]
Thanks for this @fjpanag 
Looking forward to see how to put this to use. I also wonder how this new functionality would deal with logging to an SD card when it is removed, as I would like log to ramlog and SD.

**Comment by fjpanag** [comment]
> Thanks for this @fjpanag
> Looking forward to see how to put this to use. I also wonder how this new functionality would deal with logging to an SD card when it is removed, as I would like log to ramlog and SD.

This is similar to my use case. I will use an SD card and serial output.  
For the moment I am thinking having the SD automounter to add/remove the file channel while the SD is mounted or unmounted.  
This is done in my "board" logic. I don't know how the kernel could handle this. (If only we had inotify....)

**Comment by v01d** [comment]
What I'm thinking is that maybe syslog file output should gracefully handle writing to non-accessible file, by continuing to work OK once/if the file reappears. I think this could be realated to how to perform log rotation. In our case we do not have a syslog daemon so something like that would happen externally.
Anyway, I might test this soon when I get into SD card logging.

**Comment by fjpanag** [comment]
> What I'm thinking is that maybe syslog file output should gracefully handle writing to non-accessible file, by continuing to work OK once/if the file reappears. 

This is already done.  
`syslog_dev_outputready()` will automatically open the file when it becomes available.

> I think this could be realated to how to perform log rotation.

Log rotation is coming. Although it won't be done in a daemon, rather only during adding of the file, to keep things simple and efficient.

**Pull request merged**
